qiskit/transpiler/passes/decoupling/insert_dd.py

```diff
@@ -0,0 +1,100 @@
+            raise TranspilerError('DD runs on physical circuits only')
+
+        new_dag = DAGCircuit()
+        for qreg in dag.qregs.values():
+            new_dag.add_qreg(qreg)
+        for creg in dag.cregs.values():
+            new_dag.add_creg(creg)
+
+        for nd in dag.topological_op_nodes():
+            if isinstance(nd.op, Delay):
+                qubit = nd.qargs[0]
+                pred = next(dag.predecessors(nd))
+                succ = next(dag.successors(nd))
+                if pred.type == 'in':  # discount initial delays
+                    new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                else:
+                    dd_sequence = []
+                    for g in self._dd_sequence:
+                        x = XGate()
+                        x.duration = self._durations.get(g, qubit)
+                        dd_sequence.append(x)
+                    dd_sequence_duration = sum([self._durations.get(g, qubit) for g in self._dd_sequence])
+                    slack = nd.op.duration - dd_sequence_duration
+                    if slack <= 0:     # dd doesn't fit
+                        new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                    else:
+                        if len(self._dd_sequence) == 2:
+                            begin = int(slack/4)
+                            new_dag.apply_operation_back(Delay(begin), [qubit])
+                            new_dag.apply_operation_back(dd_sequence[0], [qubit])
+                            new_dag.apply_operation_back(Delay(slack-2*begin), [qubit])
+                            new_dag.apply_operation_back(dd_sequence[1], [qubit])
+                            new_dag.apply_operation_back(Delay(begin), [qubit])
+                        elif len(self._dd_sequence) == 1: 
+                            if isinstance(succ.op, U3Gate): # only do it if the rest can absorb
+                                begin = int(slack/2)
+                                new_dag.apply_operation_back(Delay(begin), [qubit])
+                                new_dag.apply_operation_back(dd_sequence[0], [qubit])
+                                new_dag.apply_operation_back(Delay(slack-begin), [qubit])
+                                # absorb an X gate into the successor (from left in circuit)
+                                theta, phi, lam = succ.op.params
+                                succ.op.params = Optimize1qGates.compose_u3(theta, phi, lam, np.pi, 0, np.pi)
+                            else:
+                                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                        else:
+                            raise TranspilerError('whats this sequence you tryna do?')
+            else:
+                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+
+        new_dag.qubit_time_available = dag.qubit_time_available
+        new_dag.duration = dag.duration
```

Comment:
If you go the new dag approach you'll also need to copy `dag.metadata` and `dag.global_phase`. Those will be covered if you use `new_dag = dag._copy_circuit_metadata()` on L51. (it might be worth adding duration and qubit_time_available to that method or moving them to the metadata)